cmd/telemetrygen/internal/common/log.go

```diff
@@ -4,20 +4,26 @@
 package common
 
 import (
-	"fmt"
-
 	grpcZap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
 	"go.uber.org/zap"
 )
 
-// CreateLogger creates a logger for use by telemetrygen
-func CreateLogger() (*zap.Logger, error) {
-	logger, err := zap.NewDevelopment()
+var (
```

Comment:
Same as in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/26701#discussion_r1328382474, I don't know for sure (can you share the data race logs?) but I think the race is likely fixable without using `init` and globals

Reply:
```
go test -race -timeout 300s -parallel 4 --tags="" -cover ./... -covermode=atomic -args -test.gocoverdir="/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/coverage/unit"
==================
WARNING: DATA RACE
Write at 0x000001cdb4e0 by goroutine 12:
  google.golang.org/grpc/grpclog.SetLoggerV2()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/grpclog/loggerv2.go:75 +0x29a
  github.com/grpc-ecosystem/go-grpc-middleware/logging/zap.ReplaceGrpcLoggerV2WithVerbosity()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/logging/zap/grpclogger.go:65 +0x276
  github.com/grpc-ecosystem/go-grpc-middleware/logging/zap.ReplaceGrpcLoggerV2()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/logging/zap/grpclogger.go:54 +0x14b
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen/internal/common.CreateLogger()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/internal/common/log.go:19 +0x112
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen/internal/traces.Start()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/internal/traces/traces.go:30 +0x4b
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces.func2()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:54 +0x46

Previous read at 0x000001cdb4e0 by goroutine 10:
  google.golang.org/grpc/internal/grpclog.InfoDepth()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/grpclog/grpclog.go:37 +0x93
  google.golang.org/grpc/grpclog.(*componentData).InfoDepth()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/grpclog/component.go:36 +0x164
  google.golang.org/grpc/internal/channelz.AddTraceEvent()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/channelz/funcs.go:304 +0x428
  google.golang.org/grpc/internal/channelz.Info()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/channelz/logging.go:35 +0xb3
  google.golang.org/grpc.(*Server).Serve()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/server.go:855 +0x4a6
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startGRPCServer.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:93 +0xe7

Goroutine 12 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:53 +0x879
  testing.tRunner()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1629 +0x47

Goroutine 10 (running) created at:
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startGRPCServer()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:90 +0x2a4
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startProtocolServers()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:138 +0x546
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:166 +0x44
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:67 +0x9a
  sync.(*Once).doSlow()
      /opt/hostedtoolcache/go/1.20.8/x64/src/sync/once.go:74 +0x101
  sync.(*Once).Do()
      /opt/hostedtoolcache/go/1.20.8/x64/src/sync/once.go:65 +0x46
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:66 +0xe4
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:64 +0x70
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:30 +0x341
  testing.tRunner()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1629 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x000001cdb4d0 by goroutine 12:
  google.golang.org/grpc/grpclog.SetLoggerV2()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/grpclog/loggerv2.go:76 +0x2fa
  github.com/grpc-ecosystem/go-grpc-middleware/logging/zap.ReplaceGrpcLoggerV2WithVerbosity()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/logging/zap/grpclogger.go:65 +0x276
  github.com/grpc-ecosystem/go-grpc-middleware/logging/zap.ReplaceGrpcLoggerV2()
      /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/logging/zap/grpclogger.go:54 +0x14b
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen/internal/common.CreateLogger()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/internal/common/log.go:19 +0x112
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen/internal/traces.Start()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/internal/traces/traces.go:30 +0x4b
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces.func2()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:54 +0x46

Previous read at 0x000001cdb4d0 by goroutine 10:
  google.golang.org/grpc/internal/grpclog.InfoDepth()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/grpclog/grpclog.go:34 +0x44
  google.golang.org/grpc/grpclog.(*componentData).InfoDepth()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/grpclog/component.go:36 +0x164
  google.golang.org/grpc/internal/channelz.AddTraceEvent()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/channelz/funcs.go:304 +0x428
  google.golang.org/grpc/internal/channelz.Info()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/internal/channelz/logging.go:35 +0xb3
  google.golang.org/grpc.(*Server).Serve()
      /home/runner/go/pkg/mod/google.golang.org/grpc@v1.58.0/server.go:855 +0x4a6
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startGRPCServer.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:93 +0xe7

Goroutine 12 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:53 +0x879
  testing.tRunner()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1629 +0x47

Goroutine 10 (running) created at:
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startGRPCServer()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:90 +0x2a4
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).startProtocolServers()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:138 +0x546
  go.opentelemetry.io/collector/receiver/otlpreceiver.(*otlpReceiver).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver/otlpreceiver@v0.85.0/otlp.go:166 +0x44
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:67 +0x9a
  sync.(*Once).doSlow()
      /opt/hostedtoolcache/go/1.20.8/x64/src/sync/once.go:74 +0x101
  sync.(*Once).Do()
      /opt/hostedtoolcache/go/1.20.8/x64/src/sync/once.go:65 +0x46
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:66 +0xe4
  go.opentelemetry.io/collector/internal/sharedcomponent.(*SharedComponent[...]).Start()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector@v0.85.0/internal/sharedcomponent/sharedcomponent.go:64 +0x70
  github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen.TestGenerateTraces()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/main_test.go:30 +0x341
  testing.tRunner()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.8/x64/src/testing/testing.go:1629 +0x47
==================
2023-09-14T23:16:34.020Z	INFO	traces/traces.go:62	starting gRPC exporter
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:489	[core][Channel #3] Channel created	{"system": "grpc", "grpc_log": true}
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:1825	[core][Channel #3] original dial target is: "127.0.0.1:42845"	{"system": "grpc", "grpc_log": true}
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:1830	[core][Channel #3] dial target "127.0.0.1:42845" parse failed: parse "127.0.0.1:42845": first path segment in URL cannot contain colon	{"system": "grpc", "grpc_log": true}
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:1846	[core][Channel #3] fallback to scheme "passthrough"	{"system": "grpc", "grpc_log": true}
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:1854	[core][Channel #3] parsed dial target is: {URL:{Scheme:passthrough Opaque: User: Host: Path:/127.0.0.1:42845 RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}	{"system": "grpc", "grpc_log": true}
2023-09-14T23:16:34.021Z	INFO	grpc@v1.58.0/clientconn.go:1[98](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6191707478/job/16810543966#step:9:99)7	[core][Channel #3] Channel authority set to "[127](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6191707478/job/16810543966#step:9:128).0.0.1:42845"	{"system": "grpc", "grpc_log": true}
[202](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6191707478/job/16810543966#step:9:203)3-09-14T23:16:34.022Z	INFO	grpc@v1.58.0/resolver_conn_wrapper.go:246	[core][Channel #3] Resolver state updated: {
  "Addresses": [
    {
      "Addr": "127.0.0.1:42845",
      "ServerName": "",
      "Attributes": null,
      "BalancerAttributes": null,
      "Metadata": null
    }
  ],
  "Endpoints": [
    {
      "Addresses": [
        {
          "Addr": "127.0.0.1:42845",
          "ServerName": "",
          "Attributes": null,
          "BalancerAttributes": null,
          "Metadata": null
        }
      ],
      "Attributes": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} (resolver returned new addresses)	{"system": "grpc", "grpc_log": true}
```

See https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6191707478/job/16810543966

Reply:
Another way to fix the race, which also fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26659, is to drop setting the grpc logger.

Reply:
Thanks for the details! It looks like this is a data race specific to tests and would not happen when running telemetrygen by itself.

In the Collector, the solution we have used so far is open-telemetry/opentelemetry-collector/pull/5272 (basically, add a config flag to disable the gRPC logger). I think this would be the simplest approach here. Would that be okay with you? If we find a better solution we can apply it also to the Collector as a whole

Reply:
yes, why not. That'd do it.

Reply:
Done!